HACKING.md

```diff
@@ -0,0 +1,409 @@
+# Developing Salt
+
+
+So you want to contribute to the Salt project? Excellent! You can help in a
+number of ways:
+
+- Using Salt and opening bug reports.
+- Joining a [working group](https://github.com/saltstack/community).
+- Answering questions on [irc][salt on freenode], the
+  [community Slack](https://saltstackcommunity.herokuapp.com/), or the
+  [salt-users mailing list](https://groups.google.com/forum/#!forum/salt-users).
+- Fixing issues.
+
+If you'd like to update docs or fix an issue, you're going to need the Salt
+repo. The best way to contribute is using [Git](https://git-scm.com/).
+
+## Environment Setup
+
+To hack on Salt or the docs you're going to need to setup your development
+environment. If you already have a workflow that you're comfortable with, you
+can use that, but otherwise this is an opinionated guide for setting up your
+dev environment. Follow these steps and you'll end out with a functioning dev
+environment and be able to submit your first PR.
+
+This guide assumes at least a passing familiarity with
+[Git](https://git-scm.com/), and that you already have it installed and
+configured.
+
+## `pyenv`, Virtual Environments, and You
+
+We recommend [pyenv](https://github.com/pyenv/pyenv), since it allows
+installing multiple different Python versions, which is important for testing
+Salt across all the versions of Python that we support.
+
+### On Linux
+
+Install pyenv: 
+
+    git clone https://github.com/pyenv/pyenv.git ~/.pyenv
+    export PATH="$HOME/.pyenv/bin:$PATH"
+    git clone https://github.com/pyenv/pyenv-virtualenv.git $(pyenv root)/plugins/pyenv-virtualenv
+
+### On Mac
+
+Install pyenv using brew:
+
+    brew update
+    brew install pyenv
+    brew install pyenv-virtualenv
+
+---
+
+Now add pyenv to your `.bashrc`:
+
+    export PATH="$HOME/.pyenv/bin:$PATH" >> ~/.bashrc
+    pyenv init 2>> ~/.bashrc
+    pyenv virtualenv-init 2>> ~/.bashrc
+
+For other shells, see [the pyenv instructions](https://github.com/pyenv/pyenv#basic-github-checkout).
+
+Go ahead and restart your shell. Now you should be able to install a new
+version of Python:
+
+    pyenv install 3.7.0
+
+If that fails, don't panic! You're probably just missing some build
+dependencies. Check out [pyenv common build
+problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
+
+Now that you've got your version of Python installed, you can create a new
+virtual environment with this command:
+
+    pyenv virtualenv 3.7.0 salt
+
+Then activate it:
+
+    pyenv activate salt
+
+Sweet! Now you're ready to clone salt so you can start hacking away!
+
+### Cloning Salt
+
```

Comment:
**Suggestion (non-blocking, open for rebuttals):** I would recommend merging the Cloning and Forking Salt into one section and open with a paragraph like this:

Salt uses the fork and clone workflow for Git contributions. See [Using the Fork-and-Branch Git Workflow](https://blog.scottlowe.org/2015/01/27/using-fork-branch-git-workflow/) for a description of the workflow and a guide for how to implement it.